pkg/api/store/cluster/cluster_store.go

```diff
@@ -434,3 +435,55 @@ func setPrivateRegistryPasswordIfNotExists(oldData, newData map[string]interface
 	}
 	values.PutValue(newData, updatedConfig, "rancherKubernetesEngineConfig", "privateRegistries")
 }
+
+func setCloudProviderPasswordFieldsIfNotExists(oldData, newData map[string]interface{}) {
+	replaceWithOldSecretIfNotExists(oldData, newData, "openstackCloudProvider", "rancherKubernetesEngineConfig", "cloudProvider", "openstackCloudProvider", "global", "password")
+	replaceWithOldSecretIfNotExists(oldData, newData, "vsphereCloudProvider", "rancherKubernetesEngineConfig", "cloudProvider", "vsphereCloudProvider", "global", "password")
+	azureCloudProviderPasswordFields := []string{"aadClientSecret", "aadClientCertPassword"}
+	for _, secretField := range azureCloudProviderPasswordFields {
+		replaceWithOldSecretIfNotExists(oldData, newData, "azureCloudProvider", "rancherKubernetesEngineConfig", "cloudProvider", "azureCloudProvider", secretField)
+	}
+	vSphereCloudProviderConfig := values.GetValueN(newData, "rancherKubernetesEngineConfig", "cloudProvider", "vsphereCloudProvider")
+	if vSphereCloudProviderConfig == nil {
+		return
+	}
+	newVCenter := values.GetValueN(newData, "rancherKubernetesEngineConfig", "cloudProvider", "vsphereCloudProvider", "virtualCenter")
+	if newVCenter != nil {
+		oldVCenter := values.GetValueN(oldData, "rancherKubernetesEngineConfig", "cloudProvider", "vsphereCloudProvider", "virtualCenter")
+		newVCenterMap := convert.ToMapInterface(newVCenter)
+		oldVCenterMap := convert.ToMapInterface(oldVCenter)
+		updatedVCenters := make(map[string]interface{})
```

Comment:
do we need `updatedVCenters`? we don't need to put anything in L472 because we already directly updating `vCenterConfig` map

Reply:
yeah makes sense, will remove it